File: src/data/data_preprocessor.py
"""
Data Preprocessor
Additional preprocessing utilities for analysis
"""

import pandas as pd
import numpy as np
from typing import List, Optional
from ..utils.config import Config
from ..utils.logger import setup_logger
from ..utils.helpers import create_lag_features, calculate_moving_average

logger = setup_logger()

class DataPreprocessor:
    """Preprocessing utilities for time series analysis"""
    
    def __init__(self, data: pd.DataFrame):
        self.data = data.copy()
        
    def create_temporal_features(self) -> pd.DataFrame:
        """
        Create temporal features (year, quarter, month)
        
        Returns:
            DataFrame with added temporal features
        """
        logger.info("Creating temporal features")
        
        df = self.data.copy()
        
        # Extract year and quarter
        df['Year'] = df['Date'].dt.year
        df['QuarterNum'] = df['Date'].dt.quarter
        df['Month'] = df['Date'].dt.month
        
        # Create cyclical features for seasonality
        df['Quarter_sin'] = np.sin(2 * np.pi * df['QuarterNum'] / 4)
        df['Quarter_cos'] = np.cos(2 * np.pi * df['QuarterNum'] / 4)
        
        return df
    
    def add_lag_features(self, indicators: List[str], lags: List[int]) -> pd.DataFrame:
        """
        Add lagged features for specified indicators
        
        Args:
            indicators: List of indicator names
            lags: List of lag periods
            
        Returns:
            DataFrame with lagged features
        """
        logger.info(f"Adding lag features: {lags}")
        
        df = self.data.copy()
        
        for indicator in indicators:
            lag_df = create_lag_features(df[indicator], lags)
            lag_df.columns = [f'{indicator}_{col}' for col in lag_df.columns]
            df = pd.concat([df, lag_df], axis=1)
        
        return df
    
    def add_moving_averages(self, indicators: List[str], windows: List[int]) -> pd.DataFrame:
        """
        Add moving average features
        
        Args:
            indicators: List of indicator names
            windows: List of window sizes
            
        Returns:
            DataFrame with moving average features
        """
        logger.info(f"Adding moving averages: {windows}")
        
        df = self.data.copy()
        
        for indicator in indicators:
            for window in windows:
                ma = calculate_moving_average(df[indicator], window)
                df[f'{indicator}_ma{window}'] = ma
        
        return df
    
    def add_rate_of_change(self, indicators: List[str], periods: List[int] = [1, 4]) -> pd.DataFrame:
        """
        Add rate of change features
        
        Args:
            indicators: List of indicator names
            periods: List of periods for change calculation
            
        Returns:
            DataFrame with rate of change features
        """
        logger.info(f"Adding rate of change features: {periods}")
        
        df = self.data.copy()
        
        for indicator in indicators:
            for period in periods:
                change = df[indicator].pct_change(periods=period) * 100
                df[f'{indicator}_change{period}'] = change
        
        return df
    
    def normalize_indicators(self, indicators: List[str], method: str = 'zscore') -> pd.DataFrame:
        """
        Normalize indicator values
        
        Args:
            indicators: List of indicator names
            method: Normalization method ('zscore' or 'minmax')
            
        Returns:
            DataFrame with normalized indicators
        """
        logger.info(f"Normalizing indicators using {method}")
        
        df = self.data.copy()
        
        for indicator in indicators:
            if method == 'zscore':
                mean = df[indicator].mean()
                std = df[indicator].std()
                df[f'{indicator}_norm'] = (df[indicator] - mean) / std
            elif method == 'minmax':
                min_val = df[indicator].min()
                max_val = df[indicator].max()
                df[f'{indicator}_norm'] = (df[indicator] - min_val) / (max_val - min_val)
        
        return df
    
    def prepare_for_modeling(self, 
                            target_indicator: str,
                            include_lags: bool = True,
                            include_ma: bool = True,
                            include_temporal: bool = True) -> pd.DataFrame:
        """
        Prepare comprehensive feature set for modeling
        
        Args:
            target_indicator: Target indicator for prediction
            include_lags: Include lagged features
            include_ma: Include moving averages
            include_temporal: Include temporal features
            
        Returns:
            DataFrame ready for modeling
        """
        logger.info(f"Preparing data for modeling: {target_indicator}")
        
        df = self.data.copy()
        
        if include_temporal:
            df = self.create_temporal_features()
        
        if include_lags:
            df = self.add_lag_features([target_indicator], [1, 2, 4])
        
        if include_ma:
            df = self.add_moving_averages([target_indicator], [2, 4])
        
        # Add rate of change
        df = self.add_rate_of_change([target_indicator])
        
        # Drop rows with NaN values created by lagging/MA
        df = df.dropna()
        
        logger.info(f"Prepared dataset shape: {df.shape}")
        
        return df

File: labor-market-analytics/server/services/data-preprocessor.js
/**
 * Data Preprocessor Service
 * Preprocessing utilities for time series analysis
 */

class DataPreprocessor {
  constructor(data) {
    this.data = JSON.parse(JSON.stringify(data)) // Deep copy
  }

  /**
   * Create temporal features
   * @returns {Array} Data with temporal features
   */
  createTemporalFeatures() {
    console.log("[DataPreprocessor] Creating temporal features")

    return this.data.map((record) => {
      const date = new Date(record.Date)
      const year = date.getFullYear()
      const month = date.getMonth() + 1
      const quarterNum = Math.ceil(month / 3)

      // Cyclical features for seasonality
      const quarterSin = Math.sin((2 * Math.PI * quarterNum) / 4)
      const quarterCos = Math.cos((2 * Math.PI * quarterNum) / 4)

      return {
        ...record,
        Year: year,
        QuarterNum: quarterNum,
        Month: month,
        Quarter_sin: quarterSin,
        Quarter_cos: quarterCos,
      }
    })
  }

  /**
   * Add lagged features
   * @param {Array} indicators - Indicator names
   * @param {Array} lags - Lag periods
   * @returns {Array} Data with lagged features
   */
  addLagFeatures(indicators, lags) {
    console.log(`[DataPreprocessor] Adding lag features: ${lags.join(", ")}`)

    let data = JSON.parse(JSON.stringify(this.data))

    indicators.forEach((indicator) => {
      lags.forEach((lag) => {
        data = data.map((record, idx) => {
          const lagValue = idx >= lag ? data[idx - lag][indicator] : null
          return {
            ...record,
            [`${indicator}_lag${lag}`]: lagValue,
          }
        })
      })
    })

    return data
  }

  /**
   * Add moving averages
   * @param {Array} indicators - Indicator names
   * @param {Array} windows - Window sizes
   * @returns {Array} Data with moving averages
   */
  addMovingAverages(indicators, windows) {
    console.log(`[DataPreprocessor] Adding moving averages: ${windows.join(", ")}`)

    let data = JSON.parse(JSON.stringify(this.data))

    indicators.forEach((indicator) => {
      windows.forEach((window) => {
        data = data.map((record, idx) => {
          let sum = 0
          let count = 0
          for (let i = Math.max(0, idx - window + 1); i <= idx; i++) {
            sum += data[i][indicator]
            count++
          }
          return {
            ...record,
            [`${indicator}_ma${window}`]: count > 0 ? sum / count : null,
          }
        })
      })
    })

    return data
  }

  /**
   * Add rate of change features
   * @param {Array} indicators - Indicator names
   * @param {Array} periods - Periods for change calculation
   * @returns {Array} Data with rate of change
   */
  addRateOfChange(indicators, periods = [1, 4]) {
    console.log(`[DataPreprocessor] Adding rate of change: ${periods.join(", ")}`)

    let data = JSON.parse(JSON.stringify(this.data))

    indicators.forEach((indicator) => {
      periods.forEach((period) => {
        data = data.map((record, idx) => {
          if (idx < period) {
            return { ...record, [`${indicator}_change${period}`]: null }
          }
          const current = data[idx][indicator]
          const previous = data[idx - period][indicator]
          const change = ((current - previous) / previous) * 100
          return {
            ...record,
            [`${indicator}_change${period}`]: change,
          }
        })
      })
    })

    return data
  }

  /**
   * Normalize indicators
   * @param {Array} indicators - Indicator names
   * @param {string} method - 'zscore' or 'minmax'
   * @returns {Array} Data with normalized indicators
   */
  normalizeIndicators(indicators, method = "zscore") {
    console.log(`[DataPreprocessor] Normalizing using ${method}`)

    let data = JSON.parse(JSON.stringify(this.data))

    indicators.forEach((indicator) => {
      const values = this.data.map((r) => r[indicator])

      if (method === "zscore") {
        const mean = values.reduce((a, b) => a + b, 0) / values.length
        const variance = values.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / values.length
        const std = Math.sqrt(variance)

        data = data.map((record) => ({
          ...record,
          [`${indicator}_norm`]: (record[indicator] - mean) / std,
        }))
      } else if (method === "minmax") {
        const min = Math.min(...values)
        const max = Math.max(...values)

        data = data.map((record) => ({
          ...record,
          [`${indicator}_norm`]: (record[indicator] - min) / (max - min),
        }))
      }
    })

    return data
  }

  /**
   * Prepare comprehensive feature set for modeling
   * @param {string} targetIndicator - Target indicator
   * @param {Object} options - Configuration options
   * @returns {Array} Data ready for modeling
   */
  prepareForModeling(targetIndicator, options = {}) {
    const {
      includeLags = true,
      includeMA = true,
      includeRateOfChange = true,
      includeNormalization = true,
      includeTemporalFeatures = true,
    } = options

    console.log(`[DataPreprocessor] Preparing data for modeling: ${targetIndicator}`)

    let data = JSON.parse(JSON.stringify(this.data))

    if (includeTemporalFeatures) {
      data = this.createTemporalFeatures()
    }

    if (includeLags) {
      data = this.addLagFeatures([targetIndicator], [1, 2, 4])
    }

    if (includeMA) {
      data = this.addMovingAverages([targetIndicator], [2, 4])
    }

    if (includeRateOfChange) {
      data = this.addRateOfChange([targetIndicator], [1, 4])
    }

    if (includeNormalization) {
      data = this.normalizeIndicators([targetIndicator], "zscore")
    }

    // Remove rows with null values
    data = data.filter((record) => Object.values(record).every((v) => v !== null && v !== undefined))

    console.log(`[DataPreprocessor] Prepared dataset shape: ${data.length} records`)

    return data
  }
}

module.exports = DataPreprocessor
